tensorflow/contrib/repeat/python/ops/repeat_op.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Python wrapper for the Repeat Op."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from tensorflow.contrib.util import loader
+from tensorflow.python.platform import resource_loader
+from tensorflow.python.framework import ops
+from tensorflow.python.framework import constant_op
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import math_ops
+
+_repeat_so = loader.load_op_library(
+    resource_loader.get_path_to_datafile("_repeat_op.so"))
+repeat = _repeat_so.repeat
+
+@ops.RegisterGradient("Repeat")
+def _RepeatGrad(op, grad):
+  """Sum reduces grads along the repeated dimension"""
+  input_raw_shape = op.inputs[0].get_shape().as_list()
+  if len(input_raw_shape) == 0:
+    input_shape = [1]
+  else:
+    input_shape = input_raw_shape
+  output_shape = op.outputs[0].get_shape().as_list()
+  repeats = op.inputs[1]
+  axis = op.get_attr("axis")
+  if axis < 0:
+    axis += len(input_shape)
+  
+  # Collapse the outer and inner dimensions
+  outer_dim = 1
+  for i in xrange(axis):
+    outer_dim *= input_shape[i]
```

Comment:
@zycdragonball is there a way to use the `repeat` operation with inputs where some dimensions are not specified in advance?  When I tested this, the operation itself works, but gradient calculation fails here with `TypeError: unsupported operand type(s) for *=: 'int' and 'NoneType'`